libsolidity/analysis/DeclarationTypeChecker.cpp

```diff
@@ -0,0 +1,296 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libsolidity/analysis/DeclarationTypeChecker.h>
+
+#include <libsolidity/analysis/ConstantEvaluator.h>
+
+#include <libsolidity/ast/TypeProvider.h>
+
+#include <liblangutil/ErrorReporter.h>
+
+#include <boost/range/adaptor/transformed.hpp>
+
+using namespace std;
+using namespace solidity::langutil;
+using namespace solidity::frontend;
+
+bool DeclarationTypeChecker::visit(ElementaryTypeName const& _typeName)
+{
+	if (!_typeName.annotation().type)
+	{
+		_typeName.annotation().type = TypeProvider::fromElementaryTypeName(_typeName.typeName());
+		if (_typeName.stateMutability().has_value())
+		{
+			// for non-address types this was already caught by the parser
+			solAssert(_typeName.annotation().type->category() == Type::Category::Address, "");
+			switch (*_typeName.stateMutability())
+			{
+				case StateMutability::Payable:
+					_typeName.annotation().type = TypeProvider::payableAddress();
+					break;
+				case StateMutability::NonPayable:
+					_typeName.annotation().type = TypeProvider::address();
+					break;
+				default:
+					m_errorReporter.typeError(
+						_typeName.location(),
+						"Address types can only be payable or non-payable."
+					);
+					break;
+			}
+		}
+	}
+	return true;
+}
+
+void DeclarationTypeChecker::endVisit(UserDefinedTypeName const& _typeName)
+{
+	Declaration const* declaration = _typeName.annotation().referencedDeclaration;
+	solAssert(declaration, "");
+
+	if (StructDefinition const* structDef = dynamic_cast<StructDefinition const*>(declaration))
+		_typeName.annotation().type = TypeProvider::structType(*structDef, DataLocation::Storage);
```

Comment:
This would change in a second step. We should first recursively walk the struct definition itself here and annotate it with a ``recursive`` flag. Then we can create the struct type with an additional ``recursive`` flag directly as constructor argument here and in ``StructDefinition::type()``.